File: 01工厂方法模式(Factory method)/factory-luffycity.py
# -*- coding:utf-8 -*-  

from abc import ABCMeta, abstractclassmethod 

class Payment(metaclass=ABCMeta):
    @abstractclassmethod
    def pay(self, money):
        pass

class Alipay(Payment):
    def __init__(self, use_huabei=False):
        self.use_huabei = use_huabei

    def pay(self, money):
        if self.use_huabei:
            print("花呗支付%d元" % money)
        else:
            print("支付宝余额支付%d元" % money)

class WechatPay(Payment):
    def pay(self, money):
        print("微信支付%d元" % money)



# # simple factory
# class PaymentFactory:
#     def create_payment(self, method):
#         if method == 'alipay':
#             return Alipay()
#         elif method == 'wechat':
#             return WechatPay()
#         elif method == 'huabei':
#             return Alipay(use_huabei=True)
#         else:
#             return TypeError("No such payment named %s" % method)
# # client
# pf = PaymentFactory()
# # p = pf.create_payment('huabei')
# p = pf.create_payment('alipay')
# p.pay(100)

# factory
class PaymentFactory(metaclass=ABCMeta):
    @abstractclassmethod 
    def create_payment(self):
        pass


class AlipayFactory(PaymentFactory):
    def create_payment(self):
        return Alipay()

class WechatPayFactory(PaymentFactory):
    def create_payment(self):
        return WechatPay()

class HuabeiFactory(PaymentFactory):
    def create_payment(self):
        return Alipay(use_huabei=True)
    
# 在工厂模式下，如果人有新需求，做如下两步
# step 1
class BankPay(Payment):
    def pay(self, money):
        print("银联支付%d元" % money)
# step 2
class BankPayFactory(PaymentFactory):
    def create_payment(self):
        return BankPay()


pf = BankPayFactory()
p = pf.create_payment()
p.pay(100)

File: 01工厂方法模式(Factory method)/factory-luffycity.go
package main

import "fmt"

type Payment interface {
	Pay(money string) string
}

type Alipay struct{}

func (*Alipay) Pay(money string) string {
	return "支付宝余额支付元:" + " " + money
}

type WechatPay struct{}

func (*WechatPay) Pay(money string) string {
	return "微信余额支付元:" + " " + money
}

type PaymentFactory interface {
	Create_payment() Payment
}

type AlipayFactory struct{}

func (ap *AlipayFactory) Create_payment() Payment {
	return &Alipay{
		// money: string,  // remain here as an example.
	}
}

type WechatPayFactory struct{}

func (wp *WechatPayFactory) Create_payment() Payment {
	return &WechatPay{}
}

func main() {
	money := "100"
	pay := &AlipayFactory{}
	// // pay := AlipayFactory()   // 不是这样调用的，会报如下错误
	//      ./factory-luffycity.go:42:22: missing argument to conversion to AlipayFactory: AlipayFactory())
	p := pay.Create_payment()
	m := p.Pay(money)
	fmt.Println(m)
}

// func main() {
// 	method := "alipay"
// 	money := "100"
// 	pay := PaymentFactory(method)
// 	m := pay.Pay(money)
// 	fmt.Println(m)
// }
